Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp

```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
 
 void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry(GridLayoutState& gridLayoutState)
 {
-    if (m_strategy->isComputingSizeContainment()) {
+    if (m_strategy->isComputingSizeContainment() || !m_grid.hasGridItems()) {
         handleInfinityGrowthLimit();
         return;
     }
 
-    Vector<GridTrack>& allTracks = tracks(m_direction);
-    Vector<GridItemWithSpan> itemsSortedByIncreasingSpan;
-    Vector<GridItemWithSpan> itemsCrossingFlexibleTracks;
-    SingleThreadWeakHashSet<RenderBox> itemsSet;
-    MasonryIndefiniteItems masonryIndefiniteItems;
+    StdMap<unsigned, MasonryMinMaxTrackSize> indefiniteTrackSizes;
+    StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItems;
+    StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItemsWithFlex;
 
-    if (m_grid.hasGridItems()) {
-        computeIndefiniteItemsForMasonry(masonryIndefiniteItems, gridLayoutState);
+    // Compute Span Sizes
+    computeIndefiniteItemsForMasonry(indefiniteTrackSizes, definiteMultiTrackItems, definiteMultiTrackItemsWithFlex, gridLayoutState);
+
+    // Single Track Spans
+    auto& allTracks = tracks(m_direction);
 
+    if (indefiniteTrackSizes.contains(1)) {
+        auto& singleTrackSpanSize = indefiniteTrackSizes.find(1)->second;
         for (auto trackIndex : m_contentSizedTracksIndex) {
-            GridIterator iterator(m_grid, m_direction, trackIndex);
-            GridTrack& track = allTracks[trackIndex];
+            auto& track = allTracks[trackIndex];
 
-            accumulateIntrinsicSizesForTrackMasonry(track, trackIndex, iterator, itemsSortedByIncreasingSpan, itemsCrossingFlexibleTracks, itemsSet, masonryIndefiniteItems, 0_lu, gridLayoutState);
+            auto itemSpan = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + 1);
+            if (spanningItemCrossesFlexibleSizedTracks(itemSpan))
+                continue;
+
+            sizeTrackToFitSingleSpanMasonryGroup(itemSpan, singleTrackSpanSize, track);
         }
-        std::sort(itemsSortedByIncreasingSpan.begin(), itemsSortedByIncreasingSpan.end());
     }
 
-    auto it = itemsSortedByIncreasingSpan.begin();
-    auto end = itemsSortedByIncreasingSpan.end();
-    while (it != end) {
-        GridItemsSpanGroupRange spanGroupRange = { it, std::upper_bound(it, end, *it) };
-        increaseSizesToAccommodateSpanningItems<TrackSizeComputationVariant::NotCrossingFlexibleTracks>(spanGroupRange, gridLayoutState);
-        it = spanGroupRange.rangeEnd;
-    }
-    GridItemsSpanGroupRange tracksGroupRange = { itemsCrossingFlexibleTracks.begin(), itemsCrossingFlexibleTracks.end() };
-    increaseSizesToAccommodateSpanningItems<TrackSizeComputationVariant::CrossingFlexibleTracks>(tracksGroupRange, gridLayoutState);
+    // Multi Tracks (w/o Flex) Spans
```

Comment:
```suggestion
    // Multi Tracks (w/o Flex) Spans
```
// Multi Tracks without Flex Spans.